generator/src/main/java/com/scottlogic/deg/generator/generation/DecisionTreeDataGenerator.java

```diff
@@ -45,37 +45,35 @@ public DecisionTreeDataGenerator(
         DecisionTree decisionTree,
         GenerationConfig generationConfig) {
 
-        final List<DecisionTree> partitionedTrees =
-            treePartitioner
-                .splitTreeIntoPartitions(decisionTree)
-                    .map(this.treeOptimiser::optimiseTree)
-                .collect(Collectors.toList());
+        monitor.generationStarting(generationConfig);
+        CombinationStrategy partitionCombiner = generationConfig.getCombinationStrategy();
 
-        final Stream<Stream<RowSpec>> rowSpecsByPartition = partitionedTrees
-            .stream()
-            .map(tree -> treeWalker.walk(tree,
-                walkerStrategyFactory.getWalkerStrategy(profile, tree, generationConfig)));
+        Stream<Stream<DataBag>> partitionedDataBags = treePartitioner
+            .splitTreeIntoPartitions(decisionTree)
+            .map(treeOptimiser::optimiseTree)
+            .map(tree -> generateForPartition(profile, tree, generationConfig));
 
-        final Stream<DataBagSource> allDataBagSources =
-            rowSpecsByPartition
-                .map(rowSpecs ->
-                    new ConcatenatingDataBagSource(
-                        rowSpecs
-                            .map(dataBagSourceFactory::createDataBagSource)));
+        return partitionCombiner.permute(partitionedDataBags)
+            .map(dataBag -> convertToGeneratedObject(profile, dataBag))
+            .limit(generationConfig.getMaxRows().orElse(GenerationConfig.Constants.DEFAULT_MAX_ROWS))
+            .peek(monitor::rowEmitted);
+    }
 
-        Stream<GeneratedObject> dataRows = new MultiplexingDataBagSource(allDataBagSources)
-            .generate(generationConfig)
-            .map(dataBag -> new GeneratedObject(
-                profile.fields.stream()
-                    .map(dataBag::getValueAndFormat)
-                    .collect(Collectors.toList()),
-                dataBag.getRowSource(profile.fields)));
+    private Stream<DataBag> generateForPartition(Profile profile, DecisionTree tree, GenerationConfig config) {
+        FixFieldStrategy fixFieldStrategy = walkerStrategyFactory.getWalkerStrategy(profile, tree, config);
 
-        monitor.generationStarting(generationConfig);
+        Stream<DataBagSource> dataBagSources = treeWalker.walk(tree, fixFieldStrategy)
+            .map(dataBagSourceFactory::createDataBagSource);
 
-        return dataRows
-            .limit(generationConfig.getMaxRows().orElse(GenerationConfig.Constants.DEFAULT_MAX_ROWS))
-            .peek(this.monitor::rowEmitted);
+        return new ConcatenatingDataBagSource(dataBagSources)
+            .generate(config);
+    }
 
+    private GeneratedObject convertToGeneratedObject(Profile profile, DataBag dataBag) {
```

Comment:
Non-essential feedback: I would do this as:

`convertToGeneratedObject(DataBag dataBag, ProfileFields fields)`

* Accept a set of fields rather than a whole profile, because that's all this method needs (present a minimal silhouette, etc etc)
* Put the arguments in the order of their essential-ness to the operation - in this case, it's the DataBag being converted, and the fields are just facilitatory context.